### DISCUSSION TITLE
[Proposal] Smart Enums In C# Like Java

### DISCUSSION THREAD

**Discussion by noumanurrahman** [dotnet/csharplang#5935] [Language Ideas] [answered]
This proposal is about `enums` which is a pretty useful type in object-oriented programming but the thing that is missing from enum in C# is a feature from java that enums can be like classes a bit like they can have special properties, and have constructors and more...

An example for it will be a simple enum `Subscription`

In Java:
```java
public enum Subscription {
    FREE("Free Tier"),
    PRIME("Premium Tier"),
    VIP("VIP Tier");

    private final String name;

    public Subscription(String name) {
        this.name = name;
    }

    public String getName() {
        return this.name;
    }
}
```

In C#:
```csharp
public enum Subscription {
    Free,
    Prime,
    Vip;
}
```

In C# I am not able to add constructors, properties, and methods. although there is a library to do this [`SmartEnum`](https://github.com/Ardalis/SmartEnum) but wouldn't be nice if we have this feature built-in to C# where we could do something like this

```csharp
public enum Subscription {
    Free("Free Tier"),
    Prime("Premium Tier"),
    Vip("VIP Tier");

    private const string Name;

    public Subscription(string bame) {
        Name = name;
    }

    public string GetName()
    {
        return Name
    }
}
```

**Comment by huoyaoyuan** [comment]
It has been discussed a lot of times. A recent discussion is #5889.

**Comment by Richiban** [comment]
The syntax isn't exactly the same, but what you're asking for can quite easily be written today with:

```cs
public class Subscription 
{
    public static readonly Subscription Free = new("Free Tier");
    public static readonly Subscription Prime = new("Premium Tier");
    public static readonly Subscription Vip = new("VIP Tier");

    public string Name { get; }

    private Subscription(string name) 
    {
        Name = name;
    }
}
```

Although you use it like an enum it's an ordinary class at this point so you're free to add whatever methods etc you like.

**Comment by svick** [comment, marked as answer]
As pointed out by @Unknown6656 in the other discussion, this is a duplicate of https://github.com/dotnet/csharplang/discussions/5937.
